src/test/ui/rfc-2497-if-let-chains/disallowed-positions.rs

```diff
@@ -157,3 +159,31 @@ fn outside_if_and_while_expr() {
     &let 0 = 0
     //~^ ERROR `let` expressions are not supported here
 }
+
+// Let's make sure that `let` inside const generic arguments are considered.
+fn inside_const_generic_arguments() {
+    struct A<const B: bool>;
+    impl<const B: bool> A<{B}> { const O: u32 = 5; }
+
+    if let A::<{
+        true && let 1 = 1 //~ ERROR `let` expressions are not supported here
+    }>::O = 5 {}
+
+    while let A::<{
+        true && let 1 = 1 //~ ERROR `let` expressions are not supported here
+    }>::O = 5 {}
+
+    if A::<{
+        true && let 1 = 1 //~ ERROR `let` expressions are not supported here
+    }>::O == 5 {}
+
+    // In the cases above we have `ExprKind::Block` to help us out.
+    // Below however, we would not have a block and so an implementation might go
+    // from visiting expressions to types without banning `let` expressions down the tree.
+    // This tests ensures that we are not caught by surprise should the parser
+    // admit non-IDENT expressions in const generic arguments.
+
+    if A::< //~ ERROR unmatched angle bracket
+        true && let 1 = 1 //~ ERROR identifiers may currently not be used for const generics
+    >::O == 5 {}
```

Comment:
It'd be nice if this recovered and actually parsed the const generic block and emitted a suggestion to surround with braces.

_Off-topic:_ I'm starting to think the `Parser` should have a `fn will_fail_parse()` method to pessimistically checkpoint and try alternative parses, as opposed to the current approach of "this might fail, checkpoint, parse, catch error, rewind, alternate parse, catch error/cancel/emit original or suggest correct code".

Reply:
cc @varkor -- and let's possibly fix this in a *different* PR. :)

Reply:
That would be nice. Feel free to open a diagnostic issue about it.

Reply:
Filed https://github.com/rust-lang/rust/issues/61175.